libvirt/domain.go

```diff
@@ -0,0 +1,171 @@
+	blockStat := make(map[string]libvirt.VirDomainBlockStats)
+	var err error
+	if len(paths) == 0 {
+		paths, err = GetDomainDisks(domain)
+		if err != nil {
+			return blockStat, err
+		}
+	}
+	for _, path := range paths {
+		stat, err := domain.BlockStats(path)
+		if err != nil {
+			return blockStat, err
+		}
+		blockStat[path] = stat
+	}
+	return blockStat, nil
+}
+
+//GetDomainInterfaces returns slice of available network devices
+func GetDomainInterfaces(domain libvirt.VirDomain) ([]string, error) {
+	var interfaces []string
+	domXML, err := getDomainXML(domain)
+	if err != nil {
+		return interfaces, err
+	}
+	if domXML.Devices.Interface == nil || domXML.Devices.Interface.Target == nil {
+		errMsg := fmt.Sprintf("Error getting interface info, from domain: %s", domXML.Name)
+		return interfaces, errors.New(errMsg)
+	}
+	interfaces = append(interfaces, domXML.Devices.Interface.Target.AttrDev)
+	return interfaces, nil
+}
+
+//GetCPUStatistics returns cpu  statistics
+func GetCPUStatistics(domain libvirt.VirDomain) (int64, error) {
+	info, err := domain.GetInfo()
+	if err != nil {
+		return 0, err
+	}
+	return int64(info.GetCpuTime()), nil
+}
+
+//GetMemoryStatistics returns Libvirt memory statistics
+func GetMemoryStatistics(domain libvirt.VirDomain, tags ...string) (map[string]int64, error) {
+	retValue := make(map[string]int64)
+	info, err := domain.MemoryStats(5, 0)
+	if err != nil {
+		return retValue, err
+	}
+
+	for _, v := range tags {
+		retValue[v], err = parseMemStats(info, memInfo.m[v])
+		if err != nil {
+			return retValue, err
+		}
+	}
+	return retValue, nil
+}
+
+//GetDomainDisks returns slice of available disk devices
+func GetDomainDisks(domain libvirt.VirDomain) ([]string, error) {
+	var disks []string
+	domXML, err := getDomainXML(domain)
+	if err != nil {
+		return disks, err
+	}
+
+	for _, disk := range domXML.Devices.Disk {
+		if disk.AttrDevice == "disk" {
+			disks = append(disks, disk.Target.AttrDev)
+		}
+	}
+
+	return disks, nil
+}
+
+func parseMemStats(memstat []libvirt.VirDomainMemoryStat, nr int32) (int64, error) {
+
+	for i := 0; i < len(memstat); i++ {
+		if memstat[i].Tag == nr {
+			return int64(memstat[i].Val), nil
+		}
+	}
+	return 0, errors.New("Error parsing MemoryInfo")
+
+}
+
+func getDomainXML(domain libvirt.VirDomain) (*Domain, error) {
+	var libvirtXML Domain
+	xmlDesc, err := domain.GetXMLDesc(0)
+	if err != nil {
+		return &Domain{}, err
+	}
+	xml.Unmarshal([]byte(xmlDesc), &libvirtXML)
+	return &libvirtXML, nil
+}
+
+// GetNovaMetadata return Nova Metadata info
+func GetNovaMetadata(domain libvirt.VirDomain) (map[string]string, error) {
+	var novaMetadata map[string]string
+	domXML, err := getDomainXML(domain)
+	if err != nil {
+		return novaMetadata, err
+	}
+
+	if domXML.Metadata.NovaInstance != nil {
+
+		novaMetadata["nova_flavor"] = domXML.Metadata.NovaInstance.NovaFlavor.AttrName
```

Comment:
at this point `novaMetadata` is `nil `, here there are assignments to `nil`, see this simple example:

```go
package main

func main() {

	var novaMetadata map[string]string
	novaMetadata["test"] = "test-string"
}
```